src/ast/passes/semantic_analyser.cpp

```diff
@@ -1132,6 +1137,26 @@ void SemanticAnalyser::visit(Call &call)
 
     call.type = CreateUInt64();
   }
+  else if (call.func == "offsetof")
+  {
+    check_nargs(call, 2);
+    auto &element_arg = *call.vargs->at(1);
+    String &element = static_cast<String&>(element_arg);
+    auto &arg = *call.vargs->at(0);
+    // check_arg() cannot be applied to offsetof(), because
+    // Type::record is first arg of offsetof(). check_arg()
+    // return true anyway.
+    if (arg.type.type != Type::record) {
```

Comment:
This comment is very confusing to me. What is the problem here? `check_arg` should do the same comparison as you do here.

Reply:
> This comment is very confusing to me. What is the problem here? `check_arg` should do the same comparison as you do here.

because `offsetof(record, x)` is different with all other `call`, `builtin`(special vargs), when `Type::record ` in `check_arg()` must be `is_final_pass()`, obviously, for `obbsetof()` `check_arg()` return `true` anyway.

Reply:
Sorry, I'm still a bit confused. IIUC, the problem is that the subsequent access to the field must be done on record types only and `check_arg` only does that check in the final pass. This problem can be solved by doing:
```cpp
else if (is_final_pass() && !call.vargs->at(0)->type.HasField(element.str))
```
below